src/spaced/spaced/game/enemy_spawner.hpp

```diff
@@ -1,19 +1,16 @@
 #pragma once
 #include <bave/graphics/particle_system.hpp>
-#include <spaced/game/enemy.hpp>
-#include <functional>
+#include <spaced/game/enemy_factory.hpp>
 
 namespace spaced {
 class EnemySpawner {
   public:
-	using Spawn = std::function<std::unique_ptr<Enemy>()>;
-
-	explicit EnemySpawner(Spawn spawn, bave::ParticleEmitter explode);
+	explicit EnemySpawner(std::unique_ptr<IEnemyFactory> factory);
```

Comment:
why the need for `EnemySpawner`, `IEnemyFactory`, and `EnemyFactoryBuilder`? what's the difference between all of these? I get that the spawner just handles the gameplay aspects and doesn't worry about how the memory is managed and stuff, but I'm more confused about the latter two.

Reply:
The spawner - which is not customizable - uses a customizable factory to drive itself. The factory builder converts JSON data into a concrete factory.